ATO-112- Add space point distortion classes

miranov25 opened the pull request:
Squashed version of  Distortion branch (Rifki Sadikin)

* JIRA discussion within ATO-112 and ATO-433
* https://alice.its.cern.ch/jira/secure/attachment/40304/40304_spaceChargeUserManual.pdf

alibuild commented:
2dc285a453c02da4c76483b1c8421af7f4c2d0b1: approval required: 1 of @jgrosseo (Jan Fiete Grosse-Oetringhaus), @chiarazampolli (Chiara Zampolli), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/540) and comment at the bottom of the page.**_

miranov25 commented:
Hallo @rsadikin @wiechula 

I squashed your changes to one commit and made pull request.
As we added new classes I 'm not bale to approve the request.
Not sure who from the list above feel to be responsible  to approve resp. review request.

#### For reviewer information:
* new distortion/correction classes were already used within TPC git
* unit  test and documentation  for  the new classes is provided:
````
       Start 105: func_TPCbase_AliTPCSpaceCharge3DDriftLine
105/105 Test #105: func_TPCbase_AliTPCSpaceCharge3DDriftLine ....   Passed   12.17 sec

```` 
Regards
Marian.

shahor02 commented on TPC/TPCbase/AliTPC3DCylindricalInterpolatorIrregular.cxx:
```diff
@@ -0,0 +1,1167 @@
+  Double_t x0 = r0 * TMath::Cos(phi0);
+  Double_t y0 = r0 * TMath::Sin(phi0);
+  Double_t x = r * TMath::Cos(phi);
+  Double_t y = r * TMath::Sin(phi);
```
x,y are never used

shahor02 commented on TPC/TPCbase/AliTPC3DCylindricalInterpolatorIrregular.cxx:
```diff
@@ -0,0 +1,1167 @@
+  if (dPhi < -TMath::Pi())
+    dPhi = TMath::TwoPi() + dPhi;
+
+  Double_t ret = (r - r0) * (r - r0) + (dPhi * dPhi) * ((r + r0) / 2.0) + (z - z0) * (z - z0);
```
I don't understand this formula, but it is certainly wrong, since the 2 term had dimension of L instead of L^2. The standard formula is r^2+r0^2 - 2 r0*r*cos(dPhi) + (z-z0)^2

rsadikin replied:
thanks for the correction, yes I should put distance formula for cylindrical coordinate. as you note.

miranov25 commented:
Hallo @rsadikin

@shahor02 - thank you for the comment. Indeed it is a bug.
@rsadikin - could you respond to Ruben comment? 
Until now we were using only regular grid. Rifki - could you prepare unit test also for the irregular grid.


Marian.

rsadikin commented:
Thanks for the review @shahor02  I will take a look again for the interpolation on irregular implementation, yes I'll prepare the unit test for irregular grid.

miranov25 commented:
Dear Rifki.

I'm sorry. I missed your comment in GIT. Concerning the commit and unit test maybe we can discuss briefly which unit test to write in order to test non uniform grid after the bug fix.
Can we discuss the content of the test.Maybe you already do something like:
* diff between the maps with uniform and non uniform grid for given input space charge
*  compare 3 granularity of the uniform (1,1/2,1/4)  with "optimal non uniform" binning

#### Question:
* how is the non uniform binning defined ?
  *  non uniform binning was never presented respect. it is not described in manual
  * is the binning specified by user or automatically 
* in case the grid is automatically adjusted which algorithm did you use ?
  * did you rebin grid automatically using gradient dE/dx of the E field?

As the  non uniform binning is still experimental I suggest to do not delay pull request and proceed further with the "TODO - exclamation mark" for non uniform binning.
Can we meet on Vidyo today - for 10 minutes discussion. 
I prefer to have the uniform part of the code  (which is working - based on the unit test) in the AliRoot.

Regards
Marian

dberzano commented:
Is this PR still in progress? Can we close it or proceed?

miranov25 commented:
Hallo Dario.

Updated version of the distortion calculation code is available. Rifki fixed problem  indicated by Ruben, and he extended unit tests.
I expect to squash changes today, tomorrow.
Unless of other opinion, I prefer to reuse this pull request - with updated code.
Is it fine, or should we close and create new pull request?

Marian

dberzano commented:
As you wish. I'd however suggest you to use this PR. If you need more time, no problem - I just wanted to check if it wasn't left open by accident :-)

miranov25 commented:
Closing the pull request and creating new pull request (https://github.com/alisw/AliRoot/pull/630) - fixing problems discussed above.
After squashing it was difficult to reuse the same pull request
